src/riot-rs-threads/src/arch/cortex_m.rs

```diff
@@ -216,53 +212,44 @@ unsafe extern "C" fn PendSV() {
 /// This function is called in PendSV.
 // TODO: make arch independent, or move to arch
 #[no_mangle]
-unsafe fn sched(old_sp: usize) -> usize {
-    let cs = CriticalSection::new();
-    let next_pid;
-
+unsafe fn sched(old_sp: usize) -> u128 {
     loop {
-        {
-            if let Some(pid) = (&*THREADS.as_ptr(cs)).runqueue.get_next() {
-                next_pid = pid;
-                break;
+        if let Some(res) = critical_section::with(|cs| {
+            let threads = &mut *THREADS.as_ptr(cs);
+            let next_pid = match threads.runqueue.get_next() {
+                Some(pid) => pid,
+                None => {
+                    cortex_m::asm::wfi();
+                    return None;
+                }
+            };
+
+            let current_pid = threads.current_pid().unwrap();
```

Comment:
This will panic once #182 lands. 

As described in #182, it's necessary for the RISC-V implementation that we can differentiate the case when the very first thread starts running. We can only do that if we don't set `current_pid` in the `start_threading` function and instead below once we actually switch to the thread. With that change, `current_pid` would be `None` the very first time this function runs.

Is there any reason why we shouldn't do it that way?

Reply:
> Is there any reason why we shouldn't do it that way?

Performance. :)

Let's get measurement in place so any arguments have number. In general, spending even one instruction extra on *every* context switch, to work around needs of the *first* context switch, should be avoided.

Heck, we carried along an idle thread for years in RIOT-c to avoid testing if there's actually a thread. :)

Reply:
But `unwrap` internally also simply uses a match to check if it's `Some` or `None`: https://doc.rust-lang.org/src/core/option.rs.html#928.

Reply:
fixed in e3966d2864a57a49628b9742e31ac3a9ad8f218c.